### PR TITLE
detect extra region requirements in impls

### LINKED ISSUE #18937: Methods in impls allowed more restrictive lifetime bounds than in the trait.
``` rust
trait Foo {
    fn foo<T>(self) -> u64;
}
impl Foo for () {
    fn foo<T: 'static>(self) -> u64 {
        //    ^^^^^^^ this should cause an error, the bound is missing from the
        // method definition in the trait, and calls are checked against that.
        std::intrinsics::TypeId::of::<&'static T>().hash()
    }
}
fn main() {
    println!("{}", ().foo::<&()>());
}
// error: internal compiler error: non-static region found when hashing a type
```

cc @nikomatsakis

### PR DESCRIPTION
**Pull request by nikomatsakis**
The current "compare method" check fails to check for the "region obligations" that accrue in the fulfillment context. This branch switches that code to create a `FnCtxt` so that it can invoke the regionck code. Previous crater runs (I haven't done one with the latest tip) have found some small number of affected crates, so I went ahead and introduced a warning cycle. I will kick off a crater run with this branch shortly.

This is a [breaking-change] because previously unsound code was accepted. The crater runs also revealed some cases where legitimate code was no longer type-checking, so the branch contains one additional (but orthogonal) change. It improves the elaborator so that we elaborate region requirements more thoroughly. In particular, if we know that `&'a T: 'b`, we now deduce that `T: 'b` and `'a: 'b`. 

I invested a certain amount of effort in getting a good error message. The error message looks like this:

```
error[E0276]: impl has stricter requirements than trait
  --> traits-elaborate-projection-region.rs:33:5
   |
21 |     fn foo() where T: 'a;
   |     --------------------- definition of `foo` from trait
...
33 |     fn foo() where U: 'a { }
   |     ^^^^^^^^^^^^^^^^^^^^^^^^ impl has extra requirement `U: 'a`
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #18937 <https://github.com/rust-lang/rust/issues/18937>
note: lint level defined here
  --> traits-elaborate-projection-region.rs:12:9
   |
12 | #![deny(extra_requirement_in_impl)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^
```

Obviously the warning only prints if this is a _new_ error (that resulted from the bugfix). But all existing errors that fit this description are updated to follow the general template. In order to get the lint to preserve the span-labels and the error code, I separate out the core `Diagnostic` type (which encapsulates the error code, message, span, and children) from the `DiagnosticBuilder` (which layers on a `Handler` that can be used to report errors). I also extended `add_lint` with an alternative `add_lint_diagnostic` that takes in a full diagnostic (cc @jonathandturner for those changes). This doesn't feel ideal but feels like it's moving in the right direction =).

r? @pnkfelix 
cc @arielb1 

Fixes #18937

### CODE REVIEW

File: src/librustc_errors/diagnostic_builder.rs [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+/// the fields and methods of the embedded `diagnostic` in a
+/// transparent way.  *However,* many of the methods are intended to
+/// be used in a chained way, and hence ought to return `self`. In
+/// that case, we can't just naively forward to the method on the
+/// `diagnostic`, because the return type would be a `&Diagnostic`
+/// instead of a `&DiagnosticBuilder<'a>`. This `forward!` macro makes
+/// it easy to declare such methods on the builder.
+macro_rules! forward {
+    // Forward pattern for &self -> &Self
+    (pub fn $n:ident(&self, $($name:ident: $ty:ty),*) -> &Self) => {
```
Comment (arielb1) [inline src/librustc_errors/diagnostic_builder.rs]: nit: not sure a non-trivial macro is better than copy-paste here.

File: src/librustc/lint/context.rs [hunk 2]
```diff
@@ -106,38 +102,22 @@ impl fmt::Debug for EarlyLint {
...
-    }
-}
-
-pub trait EarlyLintMessage {
-    fn into_diagnostic(self, span: Span) -> Diagnostic;
+pub trait IntoEarlyLint {
+    fn into_early_lint(self, id: LintId) -> EarlyLint;
 }
 
-impl EarlyLintMessage for String {
```
Comment (arielb1) [inline src/librustc/lint/context.rs]: nit: merge this and the previous commit?

File: src/librustc/traits/error_reporting.rs [hunk 3]
```diff
@@ -866,6 +866,11 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
                 err.note(&format!("required so that reference `{}` does not outlive its referent",
                                   ref_ty));
             }
+            ObligationCauseCode::ObjectTypeBound(object_ty, region) => {
```
Comment (arielb1) [inline src/librustc/traits/error_reporting.rs]: Test for this error message?

File: src/librustc/infer/mod.rs [hunk 4]
```diff
@@ -1779,6 +1792,32 @@ impl<'tcx> SubregionOrigin<'tcx> {
             AddrOf(a) => a,
             AutoBorrow(a) => a,
             SafeDestructor(a) => a,
+            CompareImplMethodObligation { span, .. } => span,
+        }
+    }
+
+    pub fn from_cause<F>(cause: &traits::ObligationCause<'tcx>,
```
Comment (arielb1) [inline src/librustc/infer/mod.rs]: nit: from_obligation_cause? *everything* is some sort of a cause.

File: src/librustc_typeck/check/compare_method.rs [hunk 5]
```diff
@@ -454,25 +469,27 @@ pub fn compare_impl_method<'a, 'tcx>(ccx: &CrateCtxt<'a, 'tcx>,
...
+            // region obligations that get overlooked.  The right
+            // thing to do is the code below. But we keep this old
+            // pass around temporarily.
+            let mut free_regions = FreeRegionMap::new();
+            free_regions.relate_free_regions_from_predicates(
+                &infcx.parameter_environment.caller_bounds);
+            infcx.resolve_regions_and_report_errors(&free_regions, impl_m_body_id);
+        } else {
+            let fcx = FnCtxt::new(&inh, tcx.types.err, impl_m_body_id);
+            fcx.regionck_item(impl_m_body_id, impl_m_span, &[]);
```
Comment (arielb1) [inline src/librustc_typeck/check/compare_method.rs]: This also calls `relate_free_regions` from the *impl*. However, the liberated LBRs here come from the *trait*, so this is quite bogus, and even if it wasn't, you must avoid cases such as
```Rust
trait Foo {
    fn foo<'a, 'b>(x: &'a &'b u32, y: &'a &'b ());
}

impl Foo for () {
    fn foo<'a, 'b>(x: &'b &'a u32, y: &'a &'b ());
}
```

Where the impl function's implied bounds tell you that `'a = 'b`, which bypasses the subtype check.

Comment (arielb1) [inline src/librustc_typeck/check/compare_method.rs]: Ignore this - I misread `regionck_item`.

**Comment by arielb1** [comment]
The type-checking section LGTM. I'll like @jonathandturner to look at the diagnostics section.

**Comment by jonathandturner** [comment]
@arielb1 Had a similar nit but that aside diagnostics looks good.

**Comment by arielb1** [comment]
Tidy is merciless
```
/build/src/librustc/infer/error_reporting.rs:334: line longer than 100 chars

/build/src/librustc_errors/diagnostic.rs: incorrect license

/build/src/librustc_errors/diagnostic_builder.rs: incorrect license
```

File: src/librustc/traits/util.rs [hunk 6]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (pnkfelix) [inline src/librustc/traits/util.rs]: delete this extraneous blank line

File: src/librustc/infer/error_reporting.rs [hunk 7]
```diff
@@ -299,44 +300,58 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
...
+                    if let SubregionOrigin::CompareImplMethodObligation { .. } = *origin {
+                        // When comparing an impl method against a
+                        // trait method, it is not helpful to suggest
+                        // changes to the impl method.  This is
+                        // because the impl method signature is being
+                        // checked using the trait's environment, so
+                        // usually the changes we suggest would
+                        // actually have to be applied to the *trait*
+                        // method (and it's not clear that the trait
+                        // method is even under the user's control).
```
Comment (pnkfelix) [inline src/librustc/infer/error_reporting.rs]: > (and it's not clear that the trait method is even under the user's control)

can we not apply some heuristics here? For example, we could detect the scenario when the trait is defined in the same crate?

File: src/librustc_errors/diagnostic_builder.rs (hunk 1 above)
Comment (pnkfelix) [inline src/librustc_errors/diagnostic_builder.rs]: (i'm fine with the macro personally, but I can see that given the simplicity of these method bodies, this battle may not be worth fighting.)

File: src/librustc_typeck/check/compare_method.rs [hunk 8]
```diff
@@ -39,8 +40,10 @@ pub fn compare_impl_method<'a, 'tcx>(ccx: &CrateCtxt<'a, 'tcx>,
                                      impl_m_body_id: ast::NodeId,
                                      trait_m: &ty::Method<'tcx>,
                                      impl_trait_ref: &ty::TraitRef<'tcx>,
-                                     trait_item_span: Option<Span>) {
-    debug!("compare_impl_method(impl_trait_ref={:?})", impl_trait_ref);
+                                     trait_item_span: Option<Span>,
+                                     old_broken_mode: bool) {
+    debug!("compare_impl_method(impl_trait_ref={:?})",
+           impl_trait_ref);
```
Comment (pnkfelix) [inline src/librustc_typeck/check/compare_method.rs]: heh, at first I was going to say that having two `debug!` calls isn't a sin, but then I realized that @arielb1 is literally pointing out that the payload itself (`impl_trait_ref`) is duplicated in both.

File: src/test/compile-fail/issue-14853.rs [hunk 9]
```diff
@@ -20,7 +20,7 @@ struct X { data: u32 }
 
 impl Something for X {
     fn yay<T: Str>(_:Option<X>, thing: &[T]) {
-    //~^ ERROR the requirement `T: Str` appears on the impl method
+    //~^ ERROR E0276
```
Comment (pnkfelix) [inline src/test/compile-fail/issue-14853.rs]: Are all of these scenarios covered by the ui tests?

 * If so, should we just delete these compile-fail tests?
 * And if not, then I as a test reviewer personally prefer keeping the longer messages, as it gives me a better chance of divining what the original intent was from the test author.

Comment (pnkfelix) [inline src/test/compile-fail/issue-14853.rs]: (but I know that @nikomatsakis and i may simply fundamentally differ in our philosophies here.)

**Comment by pnkfelix** [comment]
These changes look fine to me; r=me assuming you address the nits pointed out by @arielb1

File: src/test/compile-fail/issue-14853.rs (hunk 9 above)
Comment (nikomatsakis) [inline src/test/compile-fail/issue-14853.rs]: I moved them to ui tests.

**Comment by arielb1** [comment]
```
---- Rust_Compiler_Error_Index_55 stdout ----
	error: only foreign functions are allowed to be variadic
 --> <anon>:8:18
  |
8 | fn foo(x: u8, ...) {}
  |                  ^

error: aborting due to previous error

thread 'Rust_Compiler_Error_Index_55' panicked at 'Box<Any>', src/librustc_errors/lib.rs:461
thread 'Rust_Compiler_Error_Index_55' panicked at 'Some expected error codes were not found: ["E0045"]', src/librustdoc/test.rs:293

```
?

**Comment by nikomatsakis** [comment]
@arielb1 wtf. That is very strange. Locally I also see a bunch of "parse-fail" test failures, which I assume is *some* kind of misconfiguration on my part.

**Comment by pnkfelix** [comment]
ping @nikomatsakis any idea what's going on with travis on this PR?

**Comment by nikomatsakis** [comment]
@pnkfelix not sure, but I think it must have to do with some of the changes I made to adjusting the error count. I've been meaning to figure this out -- hopefully I'll find some time today.

**Comment by nikomatsakis** [comment]
@pnkfelix so I think I tracked down the problem -- or at least I tracked down a problem I see locally. Locally, some of the parse-fail tests are failing. We used to print this:

```
error: a `where` clause must have at least one predicate in it
  --> /home/nmatsakis/versioned/rust-7/src/test/parse-fail/where-clauses-no-bounds-or-predicates.rs:13:36
   |
13 | fn equal1<T>(_: &T, _: &T) -> bool where {
   |                                    ^^^^^

error: each predicate in a `where` clause must have at least one bound in it
  --> /home/nmatsakis/versioned/rust-7/src/test/parse-fail/where-clauses-no-bounds-or-predicates.rs:18:42
   |
18 | fn equal2<T>(_: &T, _: &T) -> bool where T: {
   |                                          ^^

error: aborting due to previous error
```

But we now print this:

```
error: a `where` clause must have at least one predicate in it
  --> /home/nmatsakis/versioned/rust-7/src/test/parse-fail/where-clauses-no-bounds-or-predicates.rs:13:36
   |
13 | fn equal1<T>(_: &T, _: &T) -> bool where {
   |                                    ^^^^^

error: aborting due to previous error
```

Note that in both cases it ends with "error: aborting due to previous error" (i.e., the singular).

I think what caused this difference is how I changed the mechanism to adjust the error count. Consider the diff for `Handler::span_err()`:

```diff
     pub fn span_err<S: Into<MultiSpan>>(&self, sp: S, msg: &str) {
         self.emit(&sp.into(), msg, Error);
-        self.bump_err_count();
         self.panic_if_treat_err_as_bug();
     }
```

I changed it so that we bump the error count in the `emit()` method -- but only if the error hasn't been canceled in the meantime. Before, we used to bump the error (in `struct_span_err`) and then, when an error was cancelled, decrement the error back down. That made my accounting go awry in the new lint code, which sometimes stores errors for later. It also seemed unnecessarily convoluted. So I changed it to just increment the error count when errors are actually emitted (so long as they haven't been canceled).

I think this also fixed a bug in the existing parser, which changes the output here. In particular, the `emit` routine also checks a flag called `continue_parse_after_error`:

```rust
    pub fn emit(&self, msp: &MultiSpan, msg: &str, lvl: Level) {
        if lvl == Warning && !self.can_emit_warnings {
            return;
        }
        let mut db = DiagnosticBuilder::new(self, lvl, msg);
        db.set_span(msp.clone());
        db.emit();
        if !self.continue_after_error.get() { // (** HERE **)
            self.abort_if_errors();
        }
    }
```

Now, in the old code, the error count had not been bumped at this point -- it got bumped after `emit()` returned. So effectively we were aborting on the 2nd error, not the first one. This is why you see the misleading error message using the singular in the old output, even though two errors are displayed.

So I would argue that this is a bugfix (cc @nrc) in the case of the parse-fail tests. I'm not sure if this explains the other tests -- I'm going to patch the reference files locally and see what happens, maybe travis is running the tests in a funny order or something.

**Comment by nikomatsakis** [comment]
Actually I'll add `-Z continue-parse-after-error`, which seems to resolve the problem.

**Comment by nikomatsakis** [comment]
This now builds locally -- or did before rebase. Let's see what travis has to say.

**Pull request merged**